compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -2772,6 +2772,51 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         Some(r)
     }
 
+    /// Make sure that we are in the condition to suggest the blanket implementation.
+    fn maybe_lint_blanket_trait_impl<T: rustc_errors::EmissionGuarantee>(
+        &self,
+        self_ty: &hir::Ty<'_>,
+        diagnostic: &mut DiagnosticBuilder<'_, T>,
+    ) {
+        let tcx = self.tcx();
+        let parent_id = tcx.hir().get_parent_item(self_ty.hir_id);
+        if let hir::Node::Item(hir::Item {
+            kind:
+                hir::ItemKind::Impl(hir::Impl {
+                    self_ty: impl_self_typ, of_trait: Some(of_trait_ref), generics, ..
+                }),
+            ..
+        }) = tcx.hir().get_by_def_id(parent_id) && self_ty.hir_id == impl_self_typ.hir_id
+        {
+            let of_trait_span = of_trait_ref.path.span;
+            // check if the trait has generics, to make a correct suggestion
+            let of_target_span = if let Some(span) = generics.span_for_param_suggestion() {
+                span
+            } else {
+                of_trait_span
+            };
+
+            let of_typ_is_local = of_trait_ref.trait_def_id().map_or(false, |def_id| def_id.is_local());
+            if of_typ_is_local {
+                let of_trait_name = tcx.sess.source_map().span_to_snippet(of_trait_span).unwrap();
+                let impl_trait_name = tcx.sess.source_map().span_to_snippet(self_ty.span).unwrap();
+                let blanket_msg = format!(
+                    "use a blanket implementation to implement {} for all types that also implement {}",
+                    of_trait_name, impl_trait_name
+                );
+                let blanket_sugg = vec![
+                    (of_target_span.shrink_to_lo(), format!("<T: {}> ", impl_trait_name)),
+                    (self_ty.span, "T".to_owned()),
+                ];
+                diagnostic.multipart_suggestion(
+                    blanket_msg,
+                    blanket_sugg,
+                    Applicability::Unspecified,
+                );
+            }
+        }
```

Comment:
1. you should use [`next_type_param_name`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/error_reporting/suggestions/trait.NextTypeParamName.html#tymethod.next_type_param_name) to get the correct next letter to use, not just `T`. Call that value `param_name`.
2. Depending if `span_for_param_suggestion` is `Some` or `None`, you will either be formatting the param suggestion like `, {param_name}: {impl_trait_name}` or `<{param_name}: {impl_trait_name}>`, respectively. This will add `{param_name}` to an existing param list if it exists.
3. Applicability should be `MaybeIncorrect` probably.
4. You should not use `unwrap` in `span_to_snippet`, or else we will panic if sourcemap can't render the span.
5. Maybe add "alternatively," (or "consider [using]") to the beginning of the message instead?

Reply:
Sorry, I left this out. If `span_for_param_suggestion` is `None`, then just use `generics.span` instead of `of_trait_span`.